### ISSUE TITLE
생성 모델 선정을 위한 BART 모델 학습

### ISSUE THREAD

**Issue by donggrii** [boostcampaitech4lv23nlp2/final-project-level3-nlp-10#13]
## 🚀 Feature Request

- 생성 모델 선정을 위한 BART 모델 학습
- [x] gogamza/kobart-base-v2
- [x] hyunwoongko/kobart
- [x] alaggung/bart-r3f

## 📎 Additional context

- https://github.com/SKT-AI/KoBART
- https://huggingface.co/hyunwoongko/kobart
- https://huggingface.co/gogamza/kobart-base-v2
- https://huggingface.co/alaggung/bart-r3f

**Comment by Jlnus** [comment]
## 🚀 Feature Request


pretraining된 [alaggung bart-r3f](https://github.com/cosmoquester/2021-dialogue-summary-competition) 모델을 사용하여 학습 시킴

### 전처리
passage 전처리 issue #10 에서 사용 한 load_pickle_data 이용

### 학습 결과
![image](https://user-images.githubusercontent.com/69241185/215725778-1e965b33-ed41-4454-9688-b5dc256bfdfa.png)

1) base_2e-4 : 기존 통합 데이터 이용 
2) remove_minite_3e-5 : minute 데이터를 제거하여 학습
3) remove_minite_1e-4 : learning rate가 너무 작아 accuracy가 느리게 증가하여 조정
4) preprocess_2e-4 : report-speech 데이터에 문제가 있어서 전처리 수정 후 학습

### rouge 점수 비교
![image](https://user-images.githubusercontent.com/69241185/215726542-9ba83bd8-51f2-4f8b-a0a1-4a9d5bc361f6.png)

- 전처리된 데이터에서 소폭 향상

### 낮은 점수 원인 분석

![image](https://user-images.githubusercontent.com/69241185/215726687-99f084f9-b5c7-4b59-b8ee-7493353970f6.png)
  
- passage의 전반적인 요약을 보여준 target summary와 달리 모델의 predict는 대부분 첫문장을 포함하며 앞 부분을 요약하였음
- alaggung/bart-r3f 모델은 일상 대화요약 데이터로 학습된 모델이며 길이가 대부분 짧지만 우리 데이터는 그렇지 않음
- 모델 input 과정에서 passage 길이가 너무 길어서 잘렸을 가능성 있음

## 📎 Additional context


https://github.com/cosmoquester/2021-dialogue-summary-competition

**Comment by donggrii** [comment]
- pre-trained된 [gogamza/kobart-base-v2](https://huggingface.co/gogamza/kobart-base-v2)모델을 사용하여 학습

### 전처리
1. `gogamza/kobart-base-v2` tokenizer를 거친 후 `input_ids` 결과를 보면 bos token, eos token이 추가되지 않아 직접 데이터에 bos token `<s>`와 eos token `</s>`를 포함시켜 적용
```python3
from transformers import AutoTokenizer

tokenizer = AutoTokenizer.from_pretrained("gogamza/kobart-base-v2")
```

> before
<img width="484" alt="image" src="https://user-images.githubusercontent.com/54230911/218295837-a360b20b-ae88-4f29-a1c5-f452386f63bb.png">
<img width="466" alt="image" src="https://user-images.githubusercontent.com/54230911/218295856-bc7898a0-e61a-4b5a-a5ac-d02be3487f0a.png">


> after
<img width="465" alt="image" src="https://user-images.githubusercontent.com/54230911/218295864-d6358ab4-3cc7-4f43-9f92-5a9f285385e8.png">
<img width="456" alt="image" src="https://user-images.githubusercontent.com/54230911/218295880-3aee31aa-2400-4369-b64a-29eeb8262977.png">


2. 대화 내 발화문을 구분하기 위해 `<sep>` token을 추가

![image](https://user-images.githubusercontent.com/54230911/218296519-f8029cb9-8d1a-41d6-985c-a8afa0067da9.png)

3. `<sep>` token이 tokenizer의 special token에 없어서 추가
> before
<img width="250" alt="image" src="https://user-images.githubusercontent.com/54230911/218296536-ff605c48-a55d-4422-b021-6c8816655c34.png">

> after
<img width="414" alt="image" src="https://user-images.githubusercontent.com/54230911/218296626-1f22ec7e-135d-405a-ac19-c0219ac54739.png">


4. tokenizer의 special token을 추가하여 모델 학습 시 embedding size가 맞지 않는 오류를 방지하기 위해 `resize_token_embeddings` 적용

```python3
model.resize_token_embeddings(tokenizer.vocab_size + added_token_num)
```

5. **중복되는 공백을 제거**하기 위해 https://github.com/boostcampaitech4lv23nlp2/final-project-level3-nlp-10/issues/10#issuecomment-1426959661 에서 사용한 전처리 활용


6. https://github.com/boostcampaitech4lv23nlp2/final-project-level3-nlp-10/issues/10#issuecomment-1426955877 에서 사용 한 `load_pickle_data` 활용

### 학습 결과



### 평가지표 점수 비교

**Comment by donggrii** [comment]
- pre-trained된 [hyunwoongko/kobart](https://huggingface.co/hyunwoongko/kobart) 모델을 사용하여 학습

### 전처리
1. `gogamza/kobart-base-v2`와 달리 `hyunwoongko/kobart` tokenizer를 거친 후 `input_ids` 결과를 보면 bos token `<s>`, eos token `</s>`가 추가됨

```python3
from transformers import PreTrainedTokenizerFast

tokenizer = PreTrainedTokenizerFast.from_pretrained("hyunwoongko/kobart")
```

![image](https://user-images.githubusercontent.com/54230911/218299658-a3fdf99e-96df-4e43-9ebd-eefb613de707.png)

![image](https://user-images.githubusercontent.com/54230911/218299668-64060a1b-b72b-499b-8589-2bf83e2fd753.png)

- 위의 `gogamza/kobart-base-v2` 모델을 사용할 때와 동일하게 적용
2. 대화 내 발화문을 구분하기 위해 `<sep>` token을 추가
3. `<sep>` token이 tokenizer의 special token에 없어서 추가
4. tokenizer의 special token을 추가하여 모델 학습 시 embedding size가 맞지 않는 오류를 방지하기 위해 resize_token_embeddings 적용
5. **중복되는 공백을 제거**하기 위해 https://github.com/boostcampaitech4lv23nlp2/final-project-level3-nlp-10/issues/10#issuecomment-1426959661 에서 사용한 전처리 활용
6. https://github.com/boostcampaitech4lv23nlp2/final-project-level3-nlp-10/issues/10#issuecomment-1426955877 에서 사용 한 `load_pickle_data` 활용

### 학습 결과

![image](https://user-images.githubusercontent.com/54230911/218300731-4e1aed1a-649a-4b1a-a232-3c005caa4536.png)


- `seed=42`로 고정하고 진행

1. hyunwoongko_kobart_v3
```yaml
max_length: 512
fp(mixed precision): True
split_by_sep_token: False
lr_scheduler: "cosine_with_restarts"
learning_rate: 3e-5
```

2. hyunwoongko_kobart_v4_nofp
    - 1과 다르게 `fp(mixed precision)`를 적용 X
```yaml
max_length: 512
fp(mixed precision): False
split_by_sep_token: False
lr_scheduler: "cosine_with_restarts"
learning_rate: 3e-5
```

3. hyunwoongko_kobart_v5_sep_data_512
    - 1과 다르게 tokenizer의 최대 길이를 512로 하고, `input_ids` 길이가 512보다 큰 데이터는 제외하고 진행
        - `train` : 65323개 중 62761개 선정 (`2562개 제외`)
        - `valid` : 7300개 중 7066개 선정 (`234개 제외`)
```yaml
max_length: 512
fp(mixed precision): True
split_by_sep_token: True
lr_scheduler: "cosine_with_restarts"
learning_rate: 3e-5
```

4. hyunwoongko_kobart_v6_nofp_sep_data_1024
    - 3과 다르게 tokenizer의 최대 길이를 1024로 하고, `input_ids` 길이가 1024보다 큰 데이터는 제외하고 진행
        - `train` : 65323개 중 64716개 선정 (`607개 제외`)
        - `valid` : 7300개 중 전부 선정 (`0개 제외`)
```yaml
max_length: 1024
fp(mixed precision): False
split_by_sep_token: True
lr_scheduler: "cosine_with_restarts"
learning_rate: 3e-5
```

5. hyunwoongko_kobart_v7_lr2e_4_sep_data_1024
    - 4와 다르게 learning_rate를 2e-4로 변경
```yaml
max_length: 1024
fp(mixed precision): True
split_by_sep_token: True
lr_scheduler: "cosine_with_restarts"
learning_rate: 2e-4
```


### 평가지표 점수 비교

![image](https://user-images.githubusercontent.com/54230911/218301129-0bd10b5d-45de-4b7c-8bd0-17d4982433ef.png)

- ROUGE-1 (F1) 점수가 가장 높은 건 (3) hyunwoongko_kobart_v5_sep_data_512
    - tokenizer의 최대 길이를 512로 하고, `input_ids` 길이가 512보다 큰 데이터는 제외하고 진행했을 때 요약 성능이 가장 좋았음
    - fp(mixed precision)을 적용했을 때 `학습 시간도 단축(8h 32m → 6h 26m)`되고 `성능도 더 좋았음(ROUGE-1: 0.2023 → 0.2075)`
    - `<sep>` token으로 대화 내 발화문을 분리했을 때, 대화 요약을 더 잘하는 것을 확인

- (3) hyunwoongko_kobart_v5_sep_data_512 & (4) hyunwoongko_kobart_v6_nofp_sep_data_1024 비교
    - 최대 길이를 1024로 했을 때 더 많은 정보를 갖고 요약할 수 있어 요약 성능이 좋을 것이라 기대했지만 오히려 평가 점수가 떨어짐 `(ROUGE-1: 0.2075 → 0.2026)`
